File: app/src/main/java/sumicya/fcmself/XposedMain.java
package sumicya.fcmself;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import sumicya.fcmself.util.FcmselfLog;
import sumicya.fcmself.xposed.AutoStartFix;
import sumicya.fcmself.xposed.BroadcastFix;
import sumicya.fcmself.xposed.KeepNotification;
import sumicya.fcmself.xposed.OplusProxyFix;
import sumicya.fcmself.xposed.ReconnectManagerFix;
import sumicya.fcmself.xposed.XposedModule;

import io.github.libxposed.api.XposedInterface;
import io.github.libxposed.api.XposedModuleInterface;

/**
 * fcmself LSPosed 入口。
 *
 * <p>入口只负责"按进程分发"，具体 Hook 哪些模块由下方的清单声明：
 * <ul>
 *   <li>{@link #SYSTEM_SERVER_MODULES}：system_server（进程身份 "android"）内安装的模块；</li>
 *   <li>{@link #PACKAGE_MODULES}：目标应用包名 -> 该进程内安装的模块。</li>
 * </ul>
 * 新增一个 Hook 模块只需在对应清单里登记一行，无需改动分发逻辑。
 *
 * <p>每个模块独立 try/catch：单个模块安装失败只影响自身，不会阻断后续模块。
 */
public class XposedMain extends io.github.libxposed.api.XposedModule {

    private static final String PKG_GMS = "com.google.android.gms";

    /** 单个 Hook 模块的构造方式。 */
    private interface ModuleFactory {
        XposedModule create(XposedInterface api, ClassLoader classLoader);
    }

    /** Hook 模块登记项（名字仅用于日志）。 */
    private static final class ModuleEntry {
        final String name;
        final ModuleFactory factory;

        ModuleEntry(String name, ModuleFactory factory) {
            this.name = name;
            this.factory = factory;
        }
    }

    /** system_server 内安装的模块，顺序即安装顺序。 */
    private static final List<ModuleEntry> SYSTEM_SERVER_MODULES = Arrays.asList(
            new ModuleEntry("BroadcastFix", BroadcastFix::new),
            new ModuleEntry("AutoStartFix", AutoStartFix::new),
            new ModuleEntry("KeepNotification", KeepNotification::new),
            new ModuleEntry("OplusProxyFix", OplusProxyFix::new));

    /** 目标进程包名 -> 该进程内安装的模块。 */
    private static final Map<String, List<ModuleEntry>> PACKAGE_MODULES = new HashMap<>();

    static {
        PACKAGE_MODULES.put(PKG_GMS, Collections.singletonList(
                new ModuleEntry("ReconnectManagerFix", ReconnectManagerFix::new)));
    }

    @Override
    public void onSystemServerStarting(XposedModuleInterface.SystemServerStartingParam param) {
        FcmselfLog.setXposed(this);
        FcmselfLog.setSelfPackageName("android");
        installAll(SYSTEM_SERVER_MODULES, param.getClassLoader());
    }

    @Override
    public void onPackageReady(XposedModuleInterface.PackageReadyParam param) {
        // 只处理清单里登记过的进程；多用户/多进程场景下只在首个包实例上安装一次
        List<ModuleEntry> modules = PACKAGE_MODULES.get(param.getPackageName());
        if (modules == null || !param.isFirstPackage()) {
            return;
        }
        FcmselfLog.setXposed(this);
        FcmselfLog.setSelfPackageName(param.getPackageName());
        installAll(modules, param.getClassLoader());
    }

    /** 逐个安装模块；单个模块失败仅记录日志，不影响其它模块。 */
    private void installAll(List<ModuleEntry> modules, ClassLoader classLoader) {
        for (ModuleEntry entry : modules) {
            try {
                entry.factory.create(this, classLoader);
            } catch (Throwable t) {
                FcmselfLog.log("模块安装失败 " + entry.name + ": " + t);
            }
        }
    }
}

File: app/src/main/kotlin/sumicya/fcmself/XposedMain.kt
package sumicya.fcmself

import io.github.libxposed.api.XposedInterface
import io.github.libxposed.api.XposedModuleInterface

import sumicya.fcmself.core.FcmselfLog
import sumicya.fcmself.mods.AutoStartFix
import sumicya.fcmself.mods.BroadcastFix
import sumicya.fcmself.mods.KeepNotification
import sumicya.fcmself.mods.MiuiLocalNotificationFix
import sumicya.fcmself.mods.OplusProxyFix
import sumicya.fcmself.mods.PowerkeeperFix
import sumicya.fcmself.mods.ReconnectManagerFix

/**
 * fcmself LSPosed 入口。
 *
 * 入口只负责「按进程分发」，具体 Hook 哪些模块由下方的清单声明：
 * - [SYSTEM_SERVER_MODULES]：system_server（进程身份 "android"）内安装的模块；
 * - [PACKAGE_MODULES]：目标应用包名 -> 该进程内安装的模块。
 *
 * 新增一个 Hook 模块只需在对应清单里登记一个构造器引用，无需改动分发逻辑。
 * 每个模块独立 try/catch：单个模块构造失败只影响自身，不会阻断后续模块
 * （模块内部各 Hook 点还有一层 [FcmselfModule.point] 的独立容错）。
 */
class XposedMain : io.github.libxposed.api.XposedModule() {

    private companion object {
        /** system_server 内安装的模块，顺序即安装顺序。 */
        val SYSTEM_SERVER_MODULES: List<(XposedInterface, ClassLoader) -> FcmselfModule> = listOf(
            ::BroadcastFix,
            ::AutoStartFix,
            ::KeepNotification,
            ::MiuiLocalNotificationFix,
            ::OplusProxyFix,
            ::PowerkeeperFix,
        )

        /** 目标进程包名 -> 该进程内安装的模块。 */
        val PACKAGE_MODULES: Map<String, List<(XposedInterface, ClassLoader) -> FcmselfModule>> = mapOf(
            "com.google.android.gms" to listOf(::ReconnectManagerFix),
        )
    }

    override fun onSystemServerStarting(param: XposedModuleInterface.SystemServerStartingParam) {
        ProcessEnv.bootstrap(this, ProcessEnv.SYSTEM_SERVER, param.classLoader)
        installAll(SYSTEM_SERVER_MODULES, param.classLoader)
    }

    override fun onPackageReady(param: XposedModuleInterface.PackageReadyParam) {
        // 只处理清单里登记过的进程；多用户/多进程场景下只在首个包实例上安装一次
        val modules = PACKAGE_MODULES[param.packageName] ?: return
        if (!param.isFirstPackage) return
        ProcessEnv.bootstrap(this, param.packageName, param.classLoader)
        installAll(modules, param.classLoader)
    }

    /** 逐个安装模块；单个模块失败仅记录日志，不影响其它模块。 */
    private fun installAll(
        modules: List<(XposedInterface, ClassLoader) -> FcmselfModule>,
        classLoader: ClassLoader,
    ) {
        for (factory in modules) {
            try {
                factory(this, classLoader)
            } catch (t: Throwable) {
                FcmselfLog.log("模块安装失败: $t")
            }
        }
    }
}
